### PR TITLE
Add Tree Group Color Shuffling

### LINKED ISSUE #4795: Shuffle tree colors within a group
Right now the colors of all trees can be shuffled with a single click. It would be nice to expose that functionality also on a tree group level.

### LINKED ISSUE #5773: Allow to set color for all trees in a group
## Detailed Description
Right-clicking a group should show a "Set color for all trees" menu entry.

## Context
When having lots of single-node-trees which semantically belong to the same cluster, it would be helpful to give all of them the same color.

### PR DESCRIPTION
**Pull request by hotzenklotz**
PR adds the functionality to set and shuffle the colors of a whole tree group. I also fixed/added some more TS type hints for that file.

### URL of deployed dev instance (used for testing):
- https://groupcolors.webknossos.xyz

### Steps to test:
- Open annotation
- Create a new group and create a few trees
- Alternatively, use this NML of single node annotations ~400 trees
[lgn_c_22_demo_synapses.nml.zip](https://github.com/scalableminds/webknossos/files/9852311/lgn_c_22_demo_synapses.nml.zip)
- Right click on any group or sub-group and play around with shuffle / set tree group colors

### Issues:
- fixes #5773
- fixes #4795

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Needs datastore update after deployment
- [x] Ready for review

### CODE REVIEW

File: frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx [hunk 1]
```diff
@@ -384,6 +435,28 @@ class TreeHierarchyView extends React.PureComponent<Props, State> {
...
+          key="shuffleTreeGroupColors"
+          onClick={() => {
+            if (id === MISSING_GROUP_ID) this.props.onShuffleAllTreeColors();
+            else this.shuffleTreeGroupColors(id);
+          }}
+          title="Shuffle Tree Colors"
+        >
+          <i className="fas fa-adjust" /> Shuffle Tree Group Colors
+        </Menu.Item>
+        {id !== MISSING_GROUP_ID ? (
```
Comment (hotzenklotz) [inline frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx:448]: The group to tree accessors/helpers do not work with the root group. Hence, we can not offer this option on the root group.

File: frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx [hunk 2]
```diff
@@ -297,14 +335,27 @@ class TreeHierarchyView extends React.PureComponent<Props, State> {
...
+
   getNodeStyleClassForBackground = (id: number) => {
     const isTreeSelected = this.props.selectedTrees.includes(id);
 
     if (isTreeSelected) {
       return "selected-tree-node";
     }
 
-    return null;
+    return undefined;
```
Comment (hotzenklotz) [inline frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx:358]: Change for type compatibility

File: frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx [hunk 3]
```diff
@@ -284,6 +300,28 @@ class TreeHierarchyView extends React.PureComponent<Props, State> {
...
+
+  setTreeGroupColor(groupId: number, color: Vector3) {
+    const groupToTreeMap = createGroupToTreesMap(this.props.trees);
+    const groupIdWithSubgroups = getGroupByIdWithSubgroups(this.props.treeGroups, groupId);
+    const shuffleTreeColorActions = groupIdWithSubgroups.flatMap((subGroupId) => {
+      if (subGroupId in groupToTreeMap)
+        return groupToTreeMap[subGroupId].map((tree) => setTreeColorAction(tree.treeId, color));
+      return [];
+    });
+    this.props.onBatchActions(shuffleTreeColorActions, "SET_TREE_COLOR");
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx:322]: `shuffleTreeColorActions` should be `setTreeColorActions`

File: frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx (hunk 1 above)
Comment (philippotto) [reply frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx:448]: For the sake of completeness/consistency, I'd add an extra method for this case (similar to the shuffle-all-trees case). I think, it should quite short. Something like:

```js
function ... {
    const setTreeColorActions = Object.keys(this.props.trees).map((treeId) => setTreeColorAction(treeId, color));
    this.props.onBatchActions(setTreeColorActions, "SET_TREE_COLOR");
}
```

**Review by philippotto — CHANGES_REQUESTED** [review]
cool stuff! props for improving the typing, too :+1: only left two minor suggestions.

File: frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx (hunk 3 above)
Comment (hotzenklotz) [reply frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx]: Done.

File: frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx (hunk 1 above)
Comment (hotzenklotz) [reply frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx]: Good idea. I added such a method.

**Review by philippotto — APPROVED** [review]
:+1: Only see my one suggestion before merging.

**Pull request merged**